### ISSUE TITLE
kvserver: pacing/admission control for mvcc gc

### ISSUE THREAD

**Issue by irfansharif** [cockroachdb/cockroach#82955]
**Is your feature request related to a problem? Please describe.**

We've seen large backlogs of GC-able keys build up when installing long-lived PTS records. When released the MVCC GC queues go full throttle GC-ing as aggressively as possible, which we've seen to be disruptive to foreground traffic.

**Describe the solution you'd like**

- [ ] Issue MVCC GC requests with the low priority bit set for admission control to deprioritize when nodes are pushed to saturation;
- [ ] Pacing knob (bytes/sec/store) if we have a large backlog of work pent up from super old protection records;
- [ ] Knob to disable MVCC GC queue cluster wide (poor man's pacing);

**Comment by mikeczabator-doordasher** [comment]
We removed a PTS that is ~50+ days old, which triggered a significant amount of GC activity.  The GC is saturating CPU to 80%+ and disks to about the same.  

![image](https://user-images.githubusercontent.com/73661528/174070144-75520f48-d9a2-435c-aca9-e3c15cde53d8.png)

<img width="808" alt="image" src="https://user-images.githubusercontent.com/73661528/174070466-1f2187be-7dac-4347-ba81-bae046cc0948.png">

**Comment by irfansharif** [comment]
If there's latency impact during bursts of GC activity, in addition to admission controlling these requests (and/or manual control knob driven pacing), https://github.com/cockroachdb/cockroach/issues/55293 will help minimize the latency impact on foreground traffic due to the latches we currently hold during GC.

**Comment by mikeczabator** [comment]
(had to remove previous post to use correct user) 

![image](https://user-images.githubusercontent.com/35580719/174081865-e06f95de-acfc-4c4a-8d1e-809d5a629cc6.png)

![image](https://user-images.githubusercontent.com/35580719/174081903-e96a80ec-eac9-4d1a-9510-a5c5799bdea9.png)

**Comment by mikeczabator** [comment]
This image and above are from 50+ days of stuck PTS being removed, thus triggering a significant amount of GC activity. 

<img width="806" alt="image" src="https://user-images.githubusercontent.com/35580719/174082188-59ad44a8-f97e-4744-9092-3204d2807875.png">

**Comment by sumeerbhola** [comment]
@mikeczabator is there a ticket associated with the previous graphs? I am curious whether this incident showed any admission control queueing (assuming v21.2+), and whether the store was overloaded (read amp), and whether the provisioned disk bandwidth was saturated.

**Comment by irfansharif** [comment]
I've lost the original thread where this issue was discussed, but during this large volume of MVCC GC since a lot of ranges got smaller in size, there was a large build up of merge queue work and subsequent activity. It's possible that the effects observed above were as a result of the merges themselves (what exactly, I'm not sure -- the snapshots, non-MVCC latch acquisition, frozen RHS). We should try to repro that independently; for this issue we should first make sure that large volumes of MVCC GC work is in fact disruptive and benefiting of admission control.

**Comment by irfansharif** [comment]
[This](https://cockroachlabs.slack.com/archives/CARM0PTK4/p1659549817733989) is an internal incident where we observed the same effects.

**Comment by sumeerbhola** [comment]
> Pacing knob (bytes/sec/store) if we have a large backlog of work pent up from super old protection records

I suggest this should be a last resort, if we find that setting a lower admission control (AC) priority and other AC improvements do not suffice.

**Comment by irfansharif** [comment]
Yup, just wanted to list there for posterity. A first step here for anything is developing a reproduction for scenarios where we suddenly have a large build up of MVCC garbage, by perhaps installing a protected timestamp record and running a form of outbox workload along side, and then dropping the record.

**Comment by irfansharif** [comment]
@aadityasondhi is going to try and take a stab during fridays. Assigning for now.

**Comment by sumeerbhola** [comment]
@bananabrick what is the state of this issue? Do we have a simple enough reproduction?

**Comment by bananabrick** [comment]
We didn't really approach this issue based on what's written in the issue description. Instead, we observed two modes of problems due to GC.

The first one was that GC could increase the size compensated scores of levels in the lsm, which would lead to L0 compactions being starved out. That problem is fixed in https://github.com/cockroachdb/cockroach/issues/104862.

The second problem is due to GC writing too much to the LSM and causing overload. We don't have a fast reproduction for that. This should help: https://github.com/cockroachdb/cockroach/commit/08cc9b17ce0a5a61745f1880fe7d627704344875. Doesn't entirely fix the problem, but the customer we saw run into this issue also had the capacity to increase compaction concurrency. For now, I vote that we table the second problem and use the compaction concurrency as a knob to alleviate the problem if it occurs in the future.

We don't have evidence of GC issues due to CPU utilization.

**Comment by sumeerbhola** [comment]
- Declaring this done wrt the write impact of GC.
- For the scan part, we are keeping the list item in the original description to integrate with elastic CPU AC , *if* we observe this to be a problem. This will require a change to the iteration in `processReplicatedKeyRange`, since the scan does not use a `BatchRequest` and directly iterates using a `storage.Snapshot`.
